File: src/main/java/com/jloom/registry/ModuleRegistry.java
package com.jloom.registry;

import org.yaml.snakeyaml.LoaderOptions;
import org.yaml.snakeyaml.Yaml;
import org.yaml.snakeyaml.constructor.SafeConstructor;

import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

public final class ModuleRegistry {

    private final Map<String, ModuleManifest> modulesById;

    private ModuleRegistry(Map<String, ModuleManifest> modulesById) {
        this.modulesById = modulesById;
    }

    public static ModuleRegistry loadBundled() {
        ManifestLoader loader = new ManifestLoader();
        Yaml yaml = new Yaml(new SafeConstructor(new LoaderOptions()));
        Map<String, ModuleManifest> byId = new LinkedHashMap<>();

        for (String indexResource : CatalogRoots.indexResources()) {
            try (InputStream indexStream = ModuleRegistry.class.getResourceAsStream(indexResource)) {
                if (indexStream == null) {
                    throw new IllegalStateException("Missing bundled catalog index: " + indexResource);
                }
                Map<String, Object> index = yaml.load(indexStream);
                @SuppressWarnings("unchecked")
                List<String> moduleIds = (List<String>) index.getOrDefault("modules", List.of());

                for (String id : moduleIds) {
                    try (InputStream manifestStream = CatalogRoots.open(id, "module.yml")) {
                        if (manifestStream == null) {
                            throw new IllegalStateException(
                                    "Listed in " + indexResource + " but missing manifest for: " + id);
                        }
                        ModuleManifest manifest = loader.load(manifestStream);
                        if (!manifest.id().equals(id)) {
                            throw new IllegalStateException(indexResource + " lists id '" + id
                                    + "' but manifest declares id '" + manifest.id() + "'");
                        }
                        if (byId.containsKey(id)) {
                            throw new IllegalStateException(
                                    "Module id '" + id + "' is declared in more than one catalog index");
                        }
                        byId.put(id, manifest);
                    }
                }
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
        }

        return new ModuleRegistry(Map.copyOf(byId));
    }

    public List<ModuleManifest> all() {
        return List.copyOf(modulesById.values());
    }

    public Optional<ModuleManifest> find(String id) {
        return Optional.ofNullable(modulesById.get(id));
    }

    public ModuleManifest require(String id) {
        return find(id).orElseThrow(() ->
                new IllegalArgumentException("No such module: '" + id + "'. Run 'jloom list' to see available modules."));
    }

    public List<ModuleManifest.Upgrade> findUpgradePath(String moduleId, String fromVersion) {
        ModuleManifest manifest = require(moduleId);
        Map<String, ModuleManifest.Upgrade> byFromVersion = new LinkedHashMap<>();
        for (ModuleManifest.Upgrade upgrade : manifest.upgrades()) {
            byFromVersion.put(upgrade.from(), upgrade);
        }

        List<ModuleManifest.Upgrade> path = new ArrayList<>();
        String current = fromVersion;
        while (!current.equals(manifest.version())) {
            ModuleManifest.Upgrade step = byFromVersion.get(current);
            if (step == null) {
                return List.of();
            }
            path.add(step);
            current = step.to();
        }
        return List.copyOf(path);
    }

    public List<String> validate(List<String> alreadyApplied, List<String> toApply) {
        List<String> problems = new ArrayList<>();

        for (int i = 0; i < toApply.size(); i++) {
            String id = toApply.get(i);
            ModuleManifest manifest = modulesById.get(id);
            if (manifest == null) {
                problems.add("Unknown module: " + id);
                continue;
            }

            List<String> satisfiedSoFar = new ArrayList<>(alreadyApplied);
            satisfiedSoFar.addAll(toApply.subList(0, i));

            for (String required : manifest.requires()) {
                if (!isSatisfied(required, satisfiedSoFar)) {
                    boolean satisfiedLaterInBatch = isSatisfied(required, toApply.subList(i + 1, toApply.size()));
                    if (satisfiedLaterInBatch) {
                        problems.add(id + " requires '" + required + "' which is in this request but listed "
                                + "AFTER " + id + " — list it earlier, e.g.: jloom add <provider>," + id);
                    } else {
                        problems.add(id + " requires '" + required + "' which is not applied and not being applied now");
                    }
                }
            }

            List<String> effective = new ArrayList<>(alreadyApplied);
            effective.addAll(toApply);
            for (String conflict : manifest.conflicts()) {
                if (effective.contains(conflict)) {
                    problems.add(id + " conflicts with '" + conflict + "', which is already applied or in this request");
                }
            }
        }
        return List.copyOf(problems);
    }

    private boolean isSatisfied(String requirement, List<String> effectiveModuleIds) {
        if (requirement.startsWith("capability:")) {
            return effectiveModuleIds.stream()
                    .map(modulesById::get)
                    .filter(Objects::nonNull)
                    .anyMatch(m -> requirement.equals(m.provides()));
        }
        return effectiveModuleIds.contains(requirement);
    }
}

File: node/src/registry/moduleRegistry.ts
import yaml from "js-yaml";
import { readFileSync } from "node:fs";
import { indexPaths, resolvePath } from "./catalogRoots.ts";
import { parseModuleManifest } from "./manifestLoader.ts";
import type { ModuleManifest, Upgrade } from "./types.ts";

export class ModuleRegistry {
  private readonly modulesById: Map<string, ModuleManifest>;

  private constructor(modulesById: Map<string, ModuleManifest>) {
    this.modulesById = modulesById;
  }

  static loadBundled(): ModuleRegistry {
    const byId = new Map<string, ModuleManifest>();

    for (const indexPath of indexPaths()) {
      const index = yaml.load(readFileSync(indexPath, "utf8")) as { modules?: string[] } | undefined;
      const moduleIds = index?.modules ?? [];

      for (const id of moduleIds) {
        const manifestPath = resolvePath(id, "module.yml");
        if (manifestPath === undefined) {
          throw new Error(`Listed in ${indexPath} but missing manifest for: ${id}`);
        }
        const raw = yaml.load(readFileSync(manifestPath, "utf8"));
        const manifest = parseModuleManifest(raw);
        if (manifest.id !== id) {
          throw new Error(`${indexPath} lists id '${id}' but manifest declares id '${manifest.id}'`);
        }
        if (byId.has(id)) {
          throw new Error(`Module id '${id}' is declared in more than one catalog index`);
        }
        byId.set(id, manifest);
      }
    }

    return new ModuleRegistry(byId);
  }

  all(): ModuleManifest[] {
    return [...this.modulesById.values()];
  }

  find(id: string): ModuleManifest | undefined {
    return this.modulesById.get(id);
  }

  require(id: string): ModuleManifest {
    const manifest = this.find(id);
    if (manifest === undefined) {
      throw new Error(`No such module: '${id}'. Run 'jloom list' to see available modules.`);
    }
    return manifest;
  }

  findUpgradePath(moduleId: string, fromVersion: string): Upgrade[] {
    const manifest = this.require(moduleId);
    const byFromVersion = new Map<string, Upgrade>();
    for (const upgrade of manifest.upgrades) {
      byFromVersion.set(upgrade.from, upgrade);
    }

    const path: Upgrade[] = [];
    let current = fromVersion;
    while (current !== manifest.version) {
      const step = byFromVersion.get(current);
      if (step === undefined) {
        return [];
      }
      path.push(step);
      current = step.to;
    }
    return path;
  }

  validate(alreadyApplied: string[], toApply: string[]): string[] {
    const problems: string[] = [];

    for (let i = 0; i < toApply.length; i++) {
      const id = toApply[i]!;
      const manifest = this.modulesById.get(id);
      if (manifest === undefined) {
        problems.push(`Unknown module: ${id}`);
        continue;
      }

      const satisfiedSoFar = [...alreadyApplied, ...toApply.slice(0, i)];
      for (const required of manifest.requires) {
        if (!this.isSatisfied(required, satisfiedSoFar)) {
          const satisfiedLaterInBatch = this.isSatisfied(required, toApply.slice(i + 1));
          if (satisfiedLaterInBatch) {
            problems.push(
              `${id} requires '${required}' which is in this request but listed AFTER ${id} — list it earlier, e.g.: jloom add <provider>,${id}`,
            );
          } else {
            problems.push(`${id} requires '${required}' which is not applied and not being applied now`);
          }
        }
      }

      const effective = [...alreadyApplied, ...toApply];
      for (const conflict of manifest.conflicts) {
        if (effective.includes(conflict)) {
          problems.push(`${id} conflicts with '${conflict}', which is already applied or in this request`);
        }
      }
    }

    return problems;
  }

  private isSatisfied(requirement: string, effectiveModuleIds: string[]): boolean {
    if (requirement.startsWith("capability:")) {
      return effectiveModuleIds.some((id) => {
        const m = this.modulesById.get(id);
        return m !== undefined && m.provides === requirement;
      });
    }
    return effectiveModuleIds.includes(requirement);
  }
}
